--- a/term_ascii.c
+++ b/term_ascii.c
@@ -51,6 +51,8 @@
 static	void		  ascii_endline(struct termp *);
 static	void		  ascii_letter(struct termp *, int);
 static	void		  ascii_setwidth(struct termp *, int, int);
+static	void		  utf8_letter(struct termp *, int);
+static	size_t		  utf8_width(const struct termp *, int);
 
 #if HAVE_WCHAR
 static	void		  locale_advance(struct termp *, size_t);
@@ -91,7 +93,7 @@
 	p->width = ascii_width;
 
 #if HAVE_WCHAR
-	if (enc != TERMENC_ASCII) {
+	if (enc == TERMENC_LOCALE) {
 
 		/*
 		 * Do not change any of this to LC_ALL.  It might break
@@ -100,9 +102,7 @@
 		 * worst case, it might even cause buffer overflows.
 		 */
 
-		v = enc == TERMENC_LOCALE ?
-		    setlocale(LC_CTYPE, "") :
-		    setlocale(LC_CTYPE, UTF8_LOCALE);
+		v = setlocale(LC_CTYPE, "");
 
 		/*
 		 * We only support UTF-8,
@@ -122,6 +122,13 @@
 		}
 	}
 #endif
+	if (enc == TERMENC_UTF8) {
+		p->enc = TERMENC_UTF8;
+		p->advance = ascii_advance;
+		p->endline = ascii_endline;
+		p->letter = utf8_letter;
+		p->width = utf8_width;
+	}
 
 	if (outopts->mdoc) {
 		p->mdocstyle = 1;
@@ -216,6 +223,22 @@
 	mant_mandoc_output_write(&byte, 1);
 }
 
+static size_t
+utf8_width(const struct termp *p, int c)
+{
+	if (c == ASCII_NBRSP)
+		c = ' ';
+	return mant_mandoc_utf8_width(c);
+}
+
+static void
+utf8_letter(struct termp *p, int c)
+{
+	if (c == ASCII_NBRSP)
+		c = ' ';
+	mant_mandoc_output_utf8(c);
+}
+
 static void
 ascii_begin(struct termp *p)
 {
