diff --git a/roff_term.c b/roff_term.c
--- a/roff_term.c
+++ b/roff_term.c
@@ -159,11 +159,10 @@ static void
 roff_term_pre_po(ROFF_TERM_ARGS)
 {
 	struct roffsu	 su;
-	static int	 po, pouse, polast;
 	int		 ponew;
 
 	/* Revert the currently active page offset. */
-	p->tcol->offset -= pouse;
+	p->tcol->offset -= p->roff_pouse;
 
 	/* Determine the requested page offset. */
 	if (n->child != NULL &&
@@ -171,18 +170,19 @@ roff_term_pre_po(ROFF_TERM_ARGS)
 		ponew = term_hen(p, &su);
 		if (*n->child->string == '+' ||
 		    *n->child->string == '-')
-			ponew += po;
+			ponew += p->roff_po;
 	} else
-		ponew = polast;
+		ponew = p->roff_polast;
 
 	/* Remeber both the previous and the newly requested offset. */
-	polast = po;
-	po = ponew;
+	p->roff_polast = p->roff_po;
+	p->roff_po = ponew;
 
 	/* Truncate to the range [-offset, 60], remember, and apply it. */
-	pouse = po >= 60 ? 60 :
-	    po < -(int)p->tcol->offset ? -(int)p->tcol->offset : po;
-	p->tcol->offset += pouse;
+	p->roff_pouse = p->roff_po >= 60 ? 60 :
+	    p->roff_po < -(int)p->tcol->offset ?
+	    -(int)p->tcol->offset : p->roff_po;
+	p->tcol->offset += p->roff_pouse;
 }
 
 static void
diff --git a/tbl_term.c b/tbl_term.c
--- a/tbl_term.c
+++ b/tbl_term.c
@@ -132,10 +132,6 @@ static  const int borders_ascii[81] = {
 	'+', '+', '+',  /* 222 */
 };
 
-/* Either of the above according to the selected output encoding. */
-static	const int *borders_locale;
-
-
 static size_t
 term_tbl_sulen(const struct roffsu *su, void *arg)
 {
@@ -163,7 +159,6 @@ term_tbl(struct termp *tp, const struct tbl_span *sp)
 {
 	const struct tbl_cell	*cp, *cpn, *cpp, *cps;
 	const struct tbl_dat	*dp;
-	static size_t		 offset;
 	size_t			 save_offset;
 	size_t			 coloff, tsz;
 	int			 hspans, ic, more;
@@ -180,7 +175,7 @@ term_tbl(struct termp *tp, const struct tbl_span *sp)
 	 */
 
 	if (tp->tbl.cols == NULL) {
-		borders_locale = tp->enc == TERMENC_UTF8 ?
+		tp->tbl_borders = tp->enc == TERMENC_UTF8 ?
 		    borders_utf8 : borders_ascii;
 
 		tp->tbl.len = term_tbl_len;
@@ -192,7 +187,7 @@ term_tbl(struct termp *tp, const struct tbl_span *sp)
 
 		/* Center the table as a whole. */
 
-		offset = tp->tcol->offset;
+		tp->tbl_offset = tp->tcol->offset;
 		if (sp->opts->opts & TBL_OPT_CENTRE) {
 			tsz = sp->opts->opts & (TBL_OPT_BOX | TBL_OPT_DBOX)
 			    ? 2 : !!sp->opts->lvert + !!sp->opts->rvert;
@@ -201,11 +196,11 @@ term_tbl(struct termp *tp, const struct tbl_span *sp)
 				    tp->tbl.cols[ic].spacing;
 			if (sp->opts->cols)
 				tsz += tp->tbl.cols[sp->opts->cols - 1].width;
-			if (offset + tsz > tp->tcol->rmargin)
+			if (tp->tbl_offset + tsz > tp->tcol->rmargin)
 				tsz -= 1;
-			offset = offset + tp->tcol->rmargin > tsz ?
-			    (offset + tp->tcol->rmargin - tsz) / 2 : 0;
-			tp->tcol->offset = offset;
+			tp->tbl_offset = tp->tbl_offset + tp->tcol->rmargin > tsz ?
+			    (tp->tbl_offset + tp->tcol->rmargin - tsz) / 2 : 0;
+			tp->tcol->offset = tp->tbl_offset;
 		}
 
 		/* Horizontal frame at the start of boxed tables. */
@@ -220,7 +215,7 @@ term_tbl(struct termp *tp, const struct tbl_span *sp)
 	/* Set up the columns. */
 
 	tp->flags |= TERMP_MULTICOL;
-	tp->tcol->offset = offset;
+	tp->tcol->offset = tp->tbl_offset;
 	horiz = 0;
 	switch (sp->pos) {
 	case TBL_SPAN_HORIZ:
@@ -788,7 +783,7 @@ tbl_fill_border(struct termp *tp, int c, size_t len)
 {
 	char	 buf[16];
 
-	if ((c = borders_locale[c]) > 127) {
+	if ((c = tp->tbl_borders[c]) > 127) {
 		(void)snprintf(buf, sizeof(buf), "\\[u%04x]", c);
 		tbl_fill_string(tp, buf, len);
 	} else
@@ -800,7 +795,7 @@ tbl_direct_border(struct termp *tp, int c, size_t len)
 {
 	size_t	 i, sz;
 
-	c = borders_locale[c];
+	c = tp->tbl_borders[c];
 	sz = (*tp->width)(tp, c);
 	for (i = 0; i < len; i += sz) {
 		(*tp->letter)(tp, c);
diff --git a/term.h b/term.h
--- a/term.h
+++ b/term.h
@@ -60,6 +60,8 @@ struct	termp_col {
 
 struct	termp {
 	struct rofftbl	  tbl;		/* Table configuration. */
+	const int	 *tbl_borders;	/* Table borders for this encoding. */
+	size_t		  tbl_offset;	/* Left offset of the current table. */
 	struct termp_col *tcols;	/* Array of table columns. */
 	struct termp_col *tcol;		/* Current table column. */
 	size_t		  maxtcol;	/* Allocated table columns. */
@@ -77,6 +79,9 @@ struct	termp {
 	int		  mdocstyle;	/* Imitate mdoc(7) output. */
 	int		  ti;		/* Temporary indent for one line. */
 	int		  skipvsp;	/* Vertical space to skip. */
+	int		  roff_po;	/* Requested page offset. */
+	int		  roff_pouse;	/* Applied page offset. */
+	int		  roff_polast;	/* Previous page offset. */
 	int		  flags;
 #define	TERMP_SENTENCE	 (1 << 0)	/* Space before a sentence. */
 #define	TERMP_NOSPACE	 (1 << 1)	/* No space before words. */
